### ISSUE TITLE
any experience on stable training of gail discriminator

### ISSUE THREAD

**Issue by xubo92** [rohitrango/BC-regularized-GAIL#3]
Hi @rohitrango 

thanks a lot for your paper and code! It helps a lot when I start to use gail and bcgail on my own dataset.

Actually recently I tried to use a similar method gaifo (gail from observation) from other code and I keep getting unstable training of discriminator (got 'nan'). I spend some time figuring out why but still not there.

Then I remember when trying your training code for gail & bc-gail, they are stable. so I was wondering if you have any experience on the good training of gail discriminator? any special things (or tricks) to be applied?

Thanks!

Best,
Xubo

**Comment by rohitrango** [comment]
Hi @xubo92 

Glad to hear that you found our work useful. 

The answer of your question depends a lot on other factors like environments, network architecture, etc. Some general pointers I can give are:
* Discriminator may be too strong - the rewards will not be very useful in that case (it may take a long time for an agent to perfectly emulate the expert actions at the states). 
* Reward bias may be another issue. On Mujoco environments, reward functions like log(D), log(D) - log(1-D) wont work due to termination bias. Only use R = -log(1-D)
* Analyse the discriminator loss, and rewards by the policy (from the discriminator, not environment). These plots can give more insights.

**Comment by xubo92** [comment]
hi @rohitrango 

thanks a lot for your quick response! Yeah, these are really helpful insights and I will go into with them and check my code.

by the way, I also noticed in your code file "gail.py", there is a function "compute_grad_pen". I guess it's used as an additional loss for training the discriminator. but I cannot find more explanation about it from your paper. could you please give me some idea of why this function is used here? is there any undesirable effect not using it? 

(ps: I think the alpha inside this function is different from the alpha for balancing the bc and rl loss in your paper).

Thanks!

**Comment by rohitrango** [comment]
This is the gradient penalty term, a popular method [1] used for regularizing the discriminator. We used some standard tricks in the paper for stable GAN training.
Yes, that alpha is a completely different value from the alpha proposed in our paper. Sorry for the confusion, I didn't check that part of the code thoroughly for the same notation.

[1] https://arxiv.org/pdf/1704.00028.pdf

**Comment by xubo92** [comment]
@rohitrango 

I see! Thank you so much for this clear explanation!
